raft/server.cc

```diff
@@ -1267,20 +1259,30 @@ future<> server_impl::enter_joint_configuration(server_address_set c_new, seasta
     if (joining.size() == 0 && leaving.size() == 0) {
         co_return;
     }
+
     _stats.add_config++;
-    co_await add_entry_internal(raft::configuration{std::move(c_new)}, wait_type::committed, as);
-}
+    const auto& e = _fsm->add_entry(raft::configuration{std::move(c_new)});
 
-future<> server_impl::set_configuration(server_address_set c_new, seastar::abort_source* as) {
-    co_await enter_joint_configuration(std::move(c_new), as);
-    // Above we co_wait that the joint configuration is committed.
+    // We've just submitted a joint configuration to be committed.
     // Immediately, without yield, once the FSM discovers
-    // this, it appends non-joint entry. Hence,
-    // at this point in execution, non-joint entry is in the log.
-    // By waiting for a follow up dummy to get committed, we
-    // automatically wait for the non-joint entry to get
-    // committed.
-    co_await add_entry_internal(log_entry::dummy(), wait_type::committed, as);
+    // joint configuration, it appends a corresponding non-joint entry.
+    // By waiting for the joint configuration first we ensure
+    // that the next non-joint configuration we get from fsm in io_fiber
+    // would be the one corresponding to our joint configuration,
+    // no matter if the leader changed in the meantime.
+
+    auto f = _conf_commit_promise.emplace().get_future();
+    try {
+        co_await wait_for_entry({.term = e.term, .idx = e.idx}, wait_type::committed, as);
+    } catch (...) {
+        // We need to 'observe' possible exceptions in f, otherwise they will be
+        // considered unhandled and cause a warning.
+        (void)f.handle_exception([id = _id] (auto e) {
+            logger.trace("[{}] error while waiting for non-joint configuration to be committed: {}", id, e);
+        });
+        throw;
+    }
+    co_await std::move(f);
```

Comment:
I think it is still possible to hang here. Suppose after appending a non joint entry in the fsm the node is disconnected. The rest of the cluster chose another leader and it commits non joint entry. This node never gets it and never resolves the future.

Also wouldn't it have been easier to get term and idx of non joint entry and simply wait for it to be committed instead of introducing the new future and an ad-hoc loop to trigger it?

Reply:
> The rest of the cluster chose another leader and it commits non joint entry. This node never gets it and never resolves the future.

But it will resolve eventually if it gets reconnected to the cluster, because it will get the nonjoint entry eventually...
Maybe not - it could get a **snapshot** that replaced this non-joint entry. So snapshots should probably resolve these futures as well (with commit status unknown exceptions probably)... With all this it does seem that we pretty much have to repeat all the special handling of `wait_for_entry`.

> Also wouldn't it have been easier to get term and idx of non joint entry and simply wait for it to be committed instead of introducing the new future and an ad-hoc loop to trigger it?

Question is how to get this (term, idx) pair.
I proposed that we extend `fsm_output` with last non-joint config entry (term, idx) or something. But is there a guarantee that this will be the entry that we're looking for and not some later one? It should be because in order to create a new config entry we must first commit the previous one, which requires IO.

Reply:
> > The rest of the cluster chose another leader and it commits non joint entry. This node never gets it and never resolves the future.
> 
> But it will resolve eventually if it gets reconnected to the cluster, because it will get the nonjoint entry eventually... 

Why would it? It is no longer part of the cluster.

>   Maybe not - it could get a **snapshot** that replaced this non-joint entry. So snapshots should probably resolve these futures as well (with commit status unknown exceptions probably)... With all this it does seem that we pretty much have to repeat all the special handling of `wait_for_entry`.
> 
> > Also wouldn't it have been easier to get term and idx of non joint entry and simply wait for it to be committed instead of introducing the new future and an ad-hoc loop to trigger it?
> 
> Question is how to get this (term, idx) pair. I proposed that we extend `fsm_output` with last non-joint config entry (term, idx) or something. But is there a guarantee that this will be the entry that we're looking for and not some later one? It should be because in order to create a new config entry we must first commit the previous one, which requires IO.

I was thinking just scanning the log here. We even have a function to get last config entry. Not sure it is race free, but in case we are still a leader it should be.

Reply:
> Why would it? It is no longer part of the cluster.

In that case I think it's not crazy to expect that the node would eventually be shut down. We should only ensure that the promise will be aborted on shutdown, which I think it is right now.

Reply:
> > Why would it? It is no longer part of the cluster.
> 
> In that case I think it's not crazy to expect that the node would eventually be shut down. We should only ensure that the promise will be aborted on shutdown, which I think it is right now.

So if we assume that the node that is out of config suppose to be shutdown what was the problem with the current code. Also I would imagine that the code that wants to shutdown a node will remove it from the config, what for the operation to complete and then shutdown. If step 2 hangs there will be no shutdown.

Reply:
> > So if we assume that the node that is out of config suppose to be shutdown what was the problem with the current code.
> 
> The problem was that a `modify_config` call in a totally healthy cluster returned an error - even though the config was modified as requested - which (IMO) was bad API. If the cluster is healthy and the call succeeds, we should complete successfully.

Fair enough.

> 
> > Also I would imagine that the code that wants to shutdown a node will remove it from the config, what for the operation to complete and then shutdown. If step 2 hangs there will be no shutdown.
> 
> Makes sense.
> 
> However, right now for the hang to occur as you say, if I understand correctly, we need a well timed network partition. My problem with the previous situation was that we hanged or threw an error without any network partitions or crashes, in a completely healthy cluster. I think it's fine if the call will hang/timeout if there's a network partition, requiring the caller to timeout&retry.

In that case I thing it would have been better to take last log entry and wait for it instead of introducing an ad-hoc future.